### PR TITLE
SALTO-2198 fetch & deploy custom records in netsuite

### PR DESCRIPTION
**Pull request by dantal4**
Using SOAP

---
_Release Notes_: 
Netsuite Adapter:
- Fetch & deploy custom records
- Add `internalId` to customlist values

---
_User Notifications_: 
Netsuite Adapter:
Use the following adapter config to fetch custom records:
```
netsuite {
  fetch = {
      include = {
        ...
        customRecords = [
          {
            name = ".*" // regex of custom record type names to fetch
            ids = [ ".*" ] // regex of custom records to fetch
          },
       ]
    }
  }
}
```
The custom record instances will look as the following:
```
netsuite.customrecord1 val_123 {
  scriptid = "val_123"
  isInactive = false
  name = "record1"
  custom_custrecord_field1 = "test"
  custom_custrecord_field2 = 15
  custom_custrecord_account = netsuite.account.instance.account1
}

```

### CODE REVIEW

File: packages/netsuite-adapter/src/custom_records.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    ))
+
+  const recordsWithNoNameErrors = recordsWithNoName.length > 0 ? [{
+    message: `Dropped custom records of type ${type.elemID.name} without 'name' attribute`,
+    severity: 'Warning' as const,
+  }] : []
+  const recordsWithDuplicateNameErrors = recordsWithDuplicateName.map(([name]) => ({
+    message: `Dropped custom records of type ${type.elemID.name} that had the same 'name' attribute: ${name}`,
+    severity: 'Warning' as const,
+  }))
```
Comment (dantal4) [inline packages/netsuite-adapter/src/custom_records.ts:64]: @tomermevorach can you help me make these error messages more accurate?

Comment (tomermevorach) [reply packages/netsuite-adapter/src/custom_records.ts:64]: 1. Custom record type '${type.elemID.name}' does not have a 'name' property, therefore, its instances cannot be fetched.
2. Found more than one instance of custom records of type '${type.elemID.name}' named '${name}'. These instances cannot be fetched, please make sure to set unique names and rerun fetch.
3. Found one or more instances of custom records of type '${type.elemID.name}' with blank names. These instances cannot be fetched, please make sure to set unique names and rerun fetch.

File: packages/netsuite-adapter/src/filters/custom_records.ts [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    [OWNER]: {
+      refType: typesMap[RECORD_REF],
+      annotations: { [CORE_ANNOTATIONS.HIDDEN_VALUE]: true },
+    },
+    [TRANSLATION_LIST]: {
+      refType: typesMap[CUSTOM_RECORD_TRANSLATION_LIST],
+    },
+  }
+
+  const getReferenceFieldType = (
```
Comment (alonstern) [inline packages/netsuite-adapter/src/filters/custom_records.ts:105]: Can you make `getReferenceFieldType` and `getFieldDef` not inner functions? I think it will make this function a bit easier to read

File: packages/netsuite-adapter/src/filters/custom_records.ts [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  type: ObjectType,
+  typesMap: Record<string, ObjectType>,
+  instancesMap: Record<string, InstanceElement[]>
+): void => {
+  const fieldNameToDef: Record<string, FieldDefinition> = {
+    [PARENT]: {
+      refType: type,
+      annotations: { isReference: true },
+    },
+    [REC_TYPE]: {
```
Comment (alonstern) [inline packages/netsuite-adapter/src/filters/custom_records.ts:92]: I a bit confused. Isn't REC_TYPE the annotation we place the reference to the type instance in? why is this a field?

**Review by alonstern — APPROVED** [review]
Comments so far, didn't go over everything yet

**Review by alonstern — DISMISSED** [review]
.

File: packages/netsuite-adapter/src/change_validators/instance_changes.ts [hunk 4]
```diff
@@ -14,11 +14,23 @@
...
 } from '@salto-io/adapter-api'
+import { isCustomRecordType } from '../types'
+
+const isCustomRecordObjectAddition = (change: Change): boolean => {
+  if (!isAdditionChange(change)) {
+    return false
+  }
+  const data = getChangeData(change)
+  return (isObjectType(data) && isCustomRecordType(data))
+    || (isField(data) && isCustomRecordType(data.parent))
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/change_validators/instance_changes.ts:28]: adding an `ObjectType` creates changes for the type and each of its fields

File: packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+const { awu } = collections.asynciterable
+
+const addFieldsToType = (type: ObjectType): void => {
+  const fieldNameToDef: Record<string, FieldDefinition> = {
+    [PARENT]: {
+      refType: type,
+      annotations: { isReference: true },
+    },
+    [REC_TYPE]: {
+      refType: new ObjectType({ elemID: new ElemID(NETSUITE, CUSTOM_RECORD_TYPE) }),
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts:32]: this hardcoded type is used on deploy. it was @ori-moisis to do it, instead of getting the type from the `elementSource`

File: packages/netsuite-adapter/src/filters/custom_records.ts (hunk 3 above)
Comment (dantal4) [reply packages/netsuite-adapter/src/filters/custom_records.ts]: removed

File: packages/netsuite-adapter/src/filters/data_instances_diff.ts [hunk 6]
```diff
@@ -32,9 +32,13 @@ export const removeIdenticalValues = async (change: ModificationChange<InstanceE
...
     change,
     async element => {
+      const includeRecType = isCustomRecordType(await element.getType())
       element.value = _.pickBy(
         element.value,
-        (_value, key) => differentKeys.has(key) || key === 'attributes'
+        (_value, key) =>
+          differentKeys.has(key)
+          || key === 'attributes'
+          || (key === 'recordType' && includeRecType)
```
Comment (alonstern) [inline packages/netsuite-adapter/src/filters/data_instances_diff.ts:41]: why is that needed? (might be worth adding a comment)

File: packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  instance.value[REC_TYPE] = type.annotations[REC_TYPE]
+}
+
+const filterCreator = (): FilterWith<'onFetch' | 'preDeploy'> => ({
+  onFetch: async elements => {
+    await awu(elements)
+      .filter(isObjectType)
+      .filter(isCustomRecordType)
+      .forEach(type => {
+        removeFieldsFromType(type)
```
Comment (alonstern) [inline packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts:77]: why you add the fields in `custom_records.ts` if you remove them here?

File: packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts:32]: Not sure I understand, why not take the type from `customTypes[CUSTOM_RECORD_TYPE]` instead of creating it again or taking it from the `elementSource`?

File: packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      })
+  },
+  preDeploy: async changes => {
+    const typeSet = new Set<string>()
+    await awu(changes)
+      .map(getChangeData)
+      .filter(isInstanceElement)
+      .map(async instance => ({ instance, type: await instance.getType() }))
+      .filter(({ type }) => isCustomRecordType(type))
+      .forEach(async ({ instance, type }) => {
```
Comment (alonstern) [inline packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts:87]: @ori-moisis Do you see a problem here with that we edit the instance's type in `preDeploy`, but not edit it back `onDeploy`?

File: packages/netsuite-adapter/src/filters/instance_references.ts [hunk 9]
```diff
@@ -74,14 +73,29 @@ const shouldExtractToGenereatedDependency = (serviceIdInfoRecord: ServiceIdInfo)
...
+    const path = instance.value[PATH]
+    return {
+      [path]: {
+        elemID: instance.elemID.createNestedID(PATH),
+        serviceID: path,
+      },
+    }
+  }
+  if (isCustomRecordType(await instance.getType(elementsSource))) {
+    const scriptId = instance.value[SCRIPT_ID]
```
Comment (alonstern) [inline packages/netsuite-adapter/src/filters/instance_references.ts:90]: are you planning to use the `SuiteQL to get the script id of all the custom record instances?

File: packages/netsuite-adapter/src/group_changes.ts [hunk 10]
```diff
@@ -179,6 +178,7 @@ const getChangeGroupIdsWithSuiteApp: ChangeGroupIdFunction = async changes => {
     { condition: isSuiteAppRecordDeletion, group: SUITEAPP_DELETING_RECORDS_GROUP_ID },
     { condition: isSuiteAppConfigChange, group: SUITEAPP_UPDATING_CONFIG_GROUP_ID },
     { condition: isSdfChange, group: SDF_CHANGE_GROUP_ID },
+    { condition: isTypeChange, group: TYPE_CHANGE_GROUP_ID },
```
Comment (alonstern) [inline packages/netsuite-adapter/src/group_changes.ts:181]: why do we need to group all the types under the same group and can't just have each one deployed separately?

File: packages/netsuite-adapter/src/query.ts [hunk 11]
```diff
@@ -53,18 +54,29 @@ export type FetchParams = {
 export const convertToQueryParams = ({ types = {}, filePaths = [] }:
   Partial<NetsuiteQueryParameters>): QueryParams => {
   const newTypes = Object.entries(types).map(([name, ids]) => ({ name, ids }))
-  return ({ types: newTypes, fileCabinet: filePaths })
+  return ({ types: newTypes, fileCabinet: filePaths, customRecords: [] })
 }
 
-export type NetsuiteQuery = {
+export type TypesQuery = {
```
Comment (alonstern) [inline packages/netsuite-adapter/src/query.ts:60]: thanks for the refactor here 👌

File: packages/netsuite-adapter/src/custom_records.ts [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+      return { elements: [], errors }
+    }
+    return { elements: [customRecordType, ...instances], errors }
+  })
+
+  return {
+    elements: results.flatMap(({ elements }) => elements)
+      .filter(element => isObjectType(element) || query.isCustomRecordMatch({
+        type: element.elemID.typeName,
+        instanceId: element.elemID.name,
```
Comment (alonstern) [inline packages/netsuite-adapter/src/custom_records.ts:152]: IIRC in the rest of the places we put the service identifier which is usually the script id here, should we do the same here?

File: packages/netsuite-adapter/src/query.ts [hunk 13]
```diff
@@ -38,6 +38,7 @@ export type FetchTypeQueryParams = {
 export type QueryParams = {
   types: FetchTypeQueryParams[]
   fileCabinet: string[]
+  customRecords: FetchTypeQueryParams[]
```
Comment (alonstern) [inline packages/netsuite-adapter/src/query.ts:41]: Don't you also need to add it in the config objectType and the `config_doc.md`

**Review by alonstern — COMMENTED** [review]
👌👌

File: packages/netsuite-adapter/src/autogen/types/custom_types/othercustomfield.ts [hunk 14]
```diff
@@ -174,7 +174,7 @@ export const othercustomfieldType = (): TypeAndInnerTypes => {
           // [CORE_ANNOTATIONS.RESTRICTION]: createRestriction({ max_length: 200 }),
         },
       }, /* Original description: This field value can be up to 200 characters long.   This field accepts references to the string custom type. */
-      rectype: {
+      recordType: {
```
Comment (omrilit) [inline packages/netsuite-adapter/src/autogen/types/custom_types/othercustomfield.ts:177]: What is this change?
We should make sure that it's aligned with the types_generator.py script since it would get overwritten upon its next

File: packages/netsuite-adapter/src/filters/custom_records.ts [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+        return typesMap[fieldTypeName]
+      }
+      const fieldTypeName = INTERNAL_ID_TO_TYPES[value.typeId].find(
+        typeName => instancesMap[typeName]?.find(
+          instance => instance.value[INTERNAL_ID] === value.internalId
+        ) !== undefined
+      )
+      return typesMap[fieldTypeName ?? RECORD_REF]
+    }
+    const customRecordType = instancesMap[CUSTOM_RECORD_TYPE]?.find(
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/filters/custom_records.ts]: I'm not 100% sure, but my guess is that there will not be a collision here, because the `internalId` of `customrecordtype` instances are used by Netsuite as `typeId`.

File: packages/netsuite-adapter/src/filters/custom_records.ts [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+    }
+    if (isRecordRefValue(value)) {
+      const fieldType = getReferenceFieldType(value)
+      return fieldType ? { refType: fieldType } : undefined
+    }
+    if (_.isArray(value)) {
+      const innerFieldDef = getFieldDef(fieldName, value[0])
+      return innerFieldDef ? { refType: new ListType(innerFieldDef.refType) } : undefined
+    }
+    return undefined
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/filters/custom_records.ts]: If there is one, there's no meaning to giving it a field type - because the field types are used to find reference by internal ids.

File: packages/netsuite-adapter/src/filters/custom_records.ts [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+const removeDuplicateRecordValues = async (
+  instances: InstanceElement[],
+  typesMap: Record<string, ObjectType>,
+  instancesMap: Record<string, InstanceElement[]>
+): Promise<void> => {
+  const newReferencesMap: Record<string, ReferenceExpression> = {}
+  instancesMap[CUSTOM_RECORD_TYPE]?.forEach(sdfInstance => {
+    const customRecordTypeName = sdfInstance.elemID.name
+    const sdfRecords = sdfInstance.value.instances?.instance
+    if (customRecordTypeName in typesMap
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/filters/custom_records.ts]: Do you think that using the scriptid will solve this?

File: packages/netsuite-adapter/src/filters/data_instances_diff.ts (hunk 6 above)
Comment (dantal4) [reply packages/netsuite-adapter/src/filters/data_instances_diff.ts:41]: it is not needed anymore

File: packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts (hunk 7 above)
Comment (dantal4) [reply packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts:77]: the field types are used to get the references in `data_instances_references.ts`

File: packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts (hunk 8 above)
Comment (dantal4) [reply packages/netsuite-adapter/src/filters/fix_custom_record_objects.ts:87]: IIUC the changes in `preDeploy` and `onDeploy' are deep clones of the original changes, separately ([here](https://github.com/salto-io/salto/blob/3027fe2f6ac0c058923296951ea59d8389ebe725/packages/netsuite-adapter/src/adapter.ts#L388-L392) and [here](https://github.com/salto-io/salto/blob/3027fe2f6ac0c058923296951ea59d8389ebe725/packages/netsuite-adapter/src/adapter.ts#L409-L414)) so any transformation of changes on `preDeploy` won't effect the changes returned from the adapter.

File: packages/netsuite-adapter/src/filters/instance_references.ts (hunk 9 above)
Comment (dantal4) [reply packages/netsuite-adapter/src/filters/instance_references.ts:90]: yes, in `filters/custom_records.ts`

File: packages/netsuite-adapter/src/group_changes.ts (hunk 10 above)
Comment (dantal4) [reply packages/netsuite-adapter/src/group_changes.ts:181]: Today the client throw if there is no group for a change ([here](https://github.com/salto-io/salto/blob/3027fe2f6ac0c058923296951ea59d8389ebe725/packages/netsuite-adapter/src/client/client.ts#L341)). We can make the client not throwing if it is a non-instance change, but IMO it's more clear this way.
@alonstern wdyt?

File: packages/netsuite-adapter/src/custom_records.ts (hunk 12 above)
Comment (dantal4) [reply packages/netsuite-adapter/src/custom_records.ts:152]: The identifier of custom records is the `name` property, because their scriptid is not a multi-env identifier (cannot be set by the user).
@alonstern Do you think we should use the 'name' property or the nacl-case elem id?

File: packages/netsuite-adapter/src/autogen/types/custom_types/othercustomfield.ts (hunk 14 above)
Comment (dantal4) [reply packages/netsuite-adapter/src/autogen/types/custom_types/othercustomfield.ts:177]: this is my mistake.

**Review by omrilit — COMMENTED** [review]
We should remember that once this one is deployed, we should instruct CS to roll it to customers by modifying the adapter config, right?

File: packages/netsuite-adapter/src/client/suiteapp_client/soap_client/soap_client.ts [hunk 18]
```diff
@@ -392,21 +404,40 @@ export default class SoapClient {
...
+      }))
+    )
+  }
+
+  private static convertToSoapTypeName(type: ObjectType, isRecordRef: boolean): string {
+    if (isRecordRef) {
+      return RECORD_REF
+    }
+    if (isCustomRecordType(type)) {
+      return 'CustomRecord'
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/client/suiteapp_client/soap_client/soap_client.ts]: moved to a const

File: packages/netsuite-adapter/src/custom_records/custom_records.ts [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+      type,
+      instances: [],
+    } : {
+      type,
+      instances: await createInstances(client, records, customRecordTypesMap[type], elemIdGetter),
+    }
+  )).toArray()
+
+  return results.flatMap(({ type, instances }) => instances.filter(
+    instance => query.isCustomRecordMatch({ type, instanceId: instance.value[SCRIPT_ID] })
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/custom_records/custom_records.ts:126]: In some accounts the records are fetched without scriptid, so in `createInstances` we complete the scriptids for the records that miss them using SuiteQL.
IMO it looks nicer to put it here but I don't have a strong opinion.

File: packages/netsuite-adapter/src/config.ts [hunk 20]
```diff
@@ -176,6 +179,9 @@ export const fetchDefault: FetchParams = {
       '^/SuiteScripts.*',
       '^/Templates.*',
     ],
+    customRecords: [{
+      name: '.*',
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/config.ts:185]: It shouldn't be too much.

**Comment by omrilit** [comment]
> We should remember that once this one is deployed, we should instruct CS to roll it to customers by modifying the adapter config, right?

Is it correct? if so, then the description  in https://github.com/salto-io/salto/pull/3199/files#diff-42d3d94b33fa6f82e7936c2a58954e14a073093e2643ce0e0901cd73aaacec95R79 is incorrect

**Comment by dantal4** [comment]
> > We should remember that once this one is deployed, we should instruct CS to roll it to customers by modifying the adapter config, right?
> 
> Is it correct? if so, then the description in https://github.com/salto-io/salto/pull/3199/files#diff-42d3d94b33fa6f82e7936c2a58954e14a073093e2643ce0e0901cd73aaacec95R79 is incorrect

Fixed doc 👌
